### PR TITLE
check if the nodes are ubuntu 16.04

### LINKED ISSUE #1657: systemctl docker restart fail on some machine
On some machine this command failed with error;
Failed to issue method call: Unit docker.service failed to load: No such file or directory. See system logs and 'systemctl status docker.service' for details.

But the command "service docker restart" can be used. We may need to fix our script to execute this command if systemctl fails.

-----------------
**Resolution**:
+ add a system validation

### PR DESCRIPTION
**Pull request by hao1939**
Add a prerequisites check for os version, as part of #1893. Fixes #1657 .

### CODE REVIEW

File: deployment/checkCmd.py [hunk 1]
```diff
@@ -2,10 +2,52 @@
...
+    regex = r"Description:\s+Ubuntu\s+16.04(.\d)?\s+LTS"
+    for host in hosts:
+        logger.info("Check os for %s ...", host["hostip"])
+        result_stdout, result_stderr = pai_common.ssh_shell_paramiko_with_result(
+            host,
+            "sudo lsb_release -d")
+        logger.info(result_stdout)
+        # should match  Ubuntu 16.04.5 LTS
+        if re.match(regex, result_stdout) is None:
+            host_not_match.append(host["hostip"])
```
Comment (scarlett2018) [inline deployment/checkCmd.py]: do you want to append a seperation charactor beween hosts?

File: deployment/checkCmd.py [hunk 2]
```diff
@@ -2,10 +2,52 @@
...
+
+    if len(host_not_match) > 0:
+        logger.error("The nodes be listed are not ubunut 16.04: %s", host_not_match)
```
Comment (scarlett2018) [inline deployment/checkCmd.py]: > "The nodes be listed are not ubunut 16.04: %s"
Maybe change to:
[Error - unsupported OS] - OpenPAI only supports ubuntu 16.04 and above OS, the following node(s) are not ubuntu 16.04: %s.

File: deployment/checkCmd.py (hunk 1 above)
Comment (hao1939) [reply deployment/checkCmd.py]: It's a list, so no need.

File: deployment/checkCmd.py (hunk 2 above)
Comment (hao1939) [reply deployment/checkCmd.py]: Will update.
One notes: we haven't tested `16.04 above`.

File: deployment/checkCmd.py [hunk 3]
```diff
@@ -2,10 +2,52 @@
...
+from k8sPaiLibrary.maintainlib import common as pai_common
 
 logger = logging.getLogger(__name__)
 
 
+def validate_config(config_path):
+    layout_file = os.path.join(config_path, "layout.yaml")
+    old_cluster_config_file = os.path.join(config_path, "cluster-configuration.yaml")
+    if(os.path.exists(old_cluster_config_file) and not os.path.exists(layout_file)):
+        logger.error("[Error], please upgrade config files!")
```
Comment (abuccts) [inline deployment/checkCmd.py]: do not need `,`

File: deployment/checkCmd.py [hunk 4]
```diff
@@ -2,10 +2,52 @@
...
+def validata_node_os(config_path):
+    layout_file = os.path.join(config_path, "layout.yaml")
+    hosts = yaml.load(open(layout_file))["machine-list"]
+    print hosts
+
+    host_not_match = []
+    regex = r"Description:\s+Ubuntu\s+16.04(.\d)?\s+LTS"
+    for host in hosts:
+        logger.info("Check os for %s ...", host["hostip"])
+        result_stdout, result_stderr = pai_common.ssh_shell_paramiko_with_result(
```
Comment (abuccts) [inline deployment/checkCmd.py]: do not need to use regex here.
you can check file `/etc/os-release`, grep `ID=ubuntu` and `VERSION_ID="16.04"`.

Comment (hao1939) [reply deployment/checkCmd.py]: Hi @abuccts, 

Agree, using grep also works. The two method are equivalent, so either is fine.

File: deployment/checkCmd.py (hunk 3 above)
Comment (hao1939) [reply deployment/checkCmd.py]: Yeah, done.

File: deployment/checkCmd.py [hunk 5]
```diff
@@ -2,10 +2,52 @@
...
 import logging.config
+import yaml
+import re
+from k8sPaiLibrary.maintainlib import common as pai_common
 
 logger = logging.getLogger(__name__)
 
 
+def validate_config(config_path):
+    layout_file = os.path.join(config_path, "layout.yaml")
```
Comment (ydye) [inline deployment/checkCmd.py]: Directly read from yaml file? Can you read the machine list or the so called layout from cluster object model? Other wise, if the name of ```machine-list``` is changed, you will have to change 2 places.

**Pull request merged**